src/layers/doubleSymLayer.jl

```diff
@@ -85,47 +82,73 @@ function initTheta(this::DoubleSymLayer{T})  where {T<:Number}
     return theta
 end
 
-function Jthetamv(this::DoubleSymLayer{T},dtheta::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
+function Jthetamv(this::DoubleSymLayer{T},dtheta::Array{T},theta::Array{T},Y::Array{T},tmp::Array{T})  where {T<:Number}
 
-    A,dA = this.activation(tmp[2],true)
     th1, th2,th3,th4    = splitWeights(this,theta)
     dth1,dth2,dth3,dth4 = splitWeights(this,dtheta)
 
+	# re-compute derivative of activation
+	KY              = copy(tmp);
+    KY,dummy,tmpNL  = apply(this.nLayer,th4,KY,true)
+    if !isempty(th2)
+     KY .+= this.Bin*th2
+    end
+    A,dA   = this.activation(KY,true)
+    
+
 	Kop    = getOp(this.K,th1)
     dKop   = getOp(this.K,dth1)
     dY     = dKop*Y
 
-    dY = Jmv(this.nLayer,dth4,dY,th4,Kop*Y,copy(tmp[1]))[2]
-    dY = dY .+ this.Bin*dth2
-
-    dY = -(Kop'*(dA.*dY) + dKop'*A) .+ this.Bout*dth3
+    dY   = Jmv(this.nLayer,dth4,dY,th4,tmp,tmpNL)[2]
+	if !isempty(dth2)
+    	dY .+= this.Bin*dth2
+	end
+    dY = -(Kop'*(dA.*dY) + dKop'*A)
+	if !isempty(dth3) 
+		dY .+= this.Bout*dth3
+	end
     return dY, dY
 end
 
-function JYmv(this::DoubleSymLayer{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
-
-    dA = this.activation(tmp[2],true)[2]
+function JYmv(this::DoubleSymLayer{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp::Array{T})  where {T<:Number}
 
     nex = div(length(dY),nFeatIn(this))
     dY  = reshape(dY,:,nex)
     Y   = reshape(Y,:,nex)
     th1, th2,th3,th4  = splitWeights(this,theta)
 
+	# re-compute derivative of activation
+	KY              = copy(tmp);
+    KY,dummy,tmpNL  = apply(this.nLayer,th4,KY,true)
+    if !isempty(th2)
+     KY .+= this.Bin*th2
+    end
+    A,dA   = this.activation(KY,true)
+    
+
     Kop = getOp(this.K,th1)
     dY = Kop*dY
-    dY = JYmv(this.nLayer,dY,th4,Kop*Y,copy(tmp[1]))[2]
+    dY = JYmv(this.nLayer,dY,th4,tmp,tmpNL)[2]
     dZ = -(Kop'*(dA.*dY))
     return dZ, dZ
 end
 
-function Jmv(this::DoubleSymLayer{T},dtheta::Array{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp)  where {T<:Number}
-    A,dA = this.activation(copy(tmp[2]),true)
+function Jmv(this::DoubleSymLayer{T},dtheta::Array{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp::Array{T})  where {T<:Number}
     nex = div(length(Y),nFeatIn(this))
 
     th1, th2,th3,th4    = splitWeights(this,theta)
     dth1,dth2,dth3,dth4 = splitWeights(this,dtheta)
 
-    Kop    = getOp(this.K,th1)
+	# re-compute derivative of activation
+	KY              = copy(tmp);
+    KY,dummy,tmpNL  = apply(this.nLayer,th4,KY,true)
```

Comment:
And we are doing it a few times.. so instead of just doing it once and storing them all we are doing it for each function!